### PR TITLE
py: Add more authorship metadata from model card

### PR DESCRIPTION
**Pull request by mofosyne**
This is also a bit of refactoring, but also would like to give users the option of adding some authorship data via model card metadata rather than via metadata override. However since Huggingface doesn't have one to one mapping, will need to add different style of ways people may want to enter such metadata organically.

- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [X] Low
  - [ ] Medium
  - [ ] High

### CODE REVIEW

File: gguf-py/gguf/metadata.py [hunk 1]
```diff
@@ -284,20 +284,64 @@ def apply_metadata_heuristic(metadata: Metadata, model_card: Optional[dict] = No
         ########################
         if model_card is not None:
 
-            if "model_name" in model_card and metadata.name is None:
-                # Not part of huggingface model card standard but notice some model creator using it
-                # such as TheBloke in 'TheBloke/Mistral-7B-Instruct-v0.2-GGUF'
-                metadata.name = model_card.get("model_name")
+            def use_model_card_metadata(metadata: Any | None, key_name: str):
+                if key_name in model_card and metadata is None:
+                    metadata = model_card.get(key_name)
```
Comment (compilade) [inline gguf-py/gguf/metadata.py:289]: Does this work? If `metadata` is `None` here, then I'm not sure changing its value would persist when exiting this function.

You'd need `metadata` to be something passed by reference (which `None` is not).

Comment (mofosyne) [reply gguf-py/gguf/metadata.py]: Ah yeah, it's an object, so will need to use getattr and setattr.

The python unit test caught it and after this fix i got all passes as per below

```
python -m unittest discover ./gguf-py -v
test_apply_metadata_heuristic_from_hf_parameters (tests.test_metadata.TestMetadataMethod) ... ok
test_apply_metadata_heuristic_from_model_card (tests.test_metadata.TestMetadataMethod) ... ok
test_apply_metadata_heuristic_from_model_dir (tests.test_metadata.TestMetadataMethod) ... ok
test_get_model_id_components (tests.test_metadata.TestMetadataMethod) ... ok
test_id_to_title (tests.test_metadata.TestMetadataMethod) ... ok
test_apply_metadata_heuristic_from_hf_parameters (tests.test_metadata.TestMetadataMethod) ... ok
test_apply_metadata_heuristic_from_model_card (tests.test_metadata.TestMetadataMethod) ... ok
test_apply_metadata_heuristic_from_model_dir (tests.test_metadata.TestMetadataMethod) ... ok
test_get_model_id_components (tests.test_metadata.TestMetadataMethod) ... ok
test_id_to_title (tests.test_metadata.TestMetadataMethod) ... ok

----------------------------------------------------------------------
Ran 10 tests in 0.004s

OK
```

Comment (mofosyne) [reply gguf-py/gguf/metadata.py]: Thanks for the catch. Hows your take on the overall concept of this PR by the way?

Comment (compilade) [reply gguf-py/gguf/metadata.py]: > Hows your take on the overall concept of this PR by the way?

I think it's great to let the model card be used for this, as long as the semantics expected by HuggingFace are not conflicting with how metadata is used here.

(For example I think HF shows warnings when there are unknown tags in the model card, or maybe I'm mistaking it for something else, but I *think* I saw a warning related to this at least once.)

I see you've put multiple variants of the names for the same metadata keys. I think this can be good for discoverability, but (eventually, when documenting this) a single form should be recommended (while still accepting the other variants).

File: gguf-py/gguf/metadata.py [hunk 2]
```diff
@@ -329,58 +376,18 @@ def apply_metadata_heuristic(metadata: Metadata, model_card: Optional[dict] = No
...
-                if metadata.languages is None:
-                    metadata.languages = []
-
-                if isinstance(language_value, str):
-                    metadata.languages.append(language_value)
-                elif isinstance(language_value, list):
-                    metadata.languages.extend(language_value)
+            use_model_card_metadata("license", "license")
+            use_model_card_metadata("license_name", "license_name")
+            use_model_card_metadata("license_link", "license_link")
```
Comment (compilade) [inline gguf-py/gguf/metadata.py:381]: What should be done when this isn't really a link?

E.g. [like used for OpenELM by Apple](https://huggingface.co/apple/OpenELM-270M-Instruct/blob/079458728ac069399c968eb2aba519b1411725ff/README.md?code=true#L4)?

Maybe this is out of scope of this PR, though.

**Review by compilade — APPROVED** [review]
The change seems good overall, but I did not test it yet.

If you've tested that it works correctly in all relevant cases, feel free to merge.

File: gguf-py/gguf/metadata.py (hunk 2 above)
Comment (mofosyne) [reply gguf-py/gguf/metadata.py:381]: ... create `license_content` ?

**Comment by mofosyne** [comment]
Had a final lookover and at least we still got the basic test cases testing that area so should be ok.

I did identify future room for improvement in that I can see people wanting to put more specific information besides an array of string IDs for base_model and datasets (e.g. name, org, version, url, etc...). But let's wait and see if people use it enough to complain about it. Since we can always add detection for dicts, when the demand arrives.
